### ISSUE TITLE
Caching issues with ReadyToRun images

### ISSUE THREAD

**Issue by ltrzesniewski** [dotnet/sdk#12379]
ReadyToRun images from a previous build can currently be reused when they should not.

Here's what contributes to the issue:

 - The `_CreateR2RImages` target compares file timestamps: as long as the input file has an earlier timestamp than the R2R file, the R2R file will not be regenerated.
 - NuGet uses the timestamps from the nupkg when restoring packages, which means files from a newly restored package will have an earlier timestamp. But changing that wouldn't help much, as a given package can already be present in the package cache.
 - The `Clean` target does not clean the R2R images

This means that you will end up with an outdated R2R image if you update a NuGet package and then republish the project, unless you delete the `obj` directory manually first.

Here's how to reproduce this:

```
dotnet new console
dotnet add package Newtonsoft.Json -v 11.0.2
dotnet publish -r win-x64 -p:PublishReadyToRun=true
dotnet add package Newtonsoft.Json -v 12.0.3
dotnet publish -r win-x64 -p:PublishReadyToRun=true
```

The `Newtonsoft.Json.dll` in the publish output will be the v11.0.2 version instead of v12.0.3. Inserting `dotnet clean -r win-x64` after the first `publish` does not help.

To recap, I see two issues:

 - R2R images should be invalidated when the source assembly is modified, using something stronger than just the file timestamp
 - The `Clean` target should delete the R2R images

**Comment by DavidZidar** [comment]
I was just hit by this issue with a nasty assembly resolution exception in runtime (in a staging environment fortunately) because ReadyToRun compilation was skipped when it shouldn't have been so an old mismatching version was published.

I am currently using .NET Core 3.1, is this fixed in .NET 5?

**Comment by AddictedCS** [comment]
Can confirm, got an old assembly linked after having the solution compiled with `/p:PublishReadyToRun`.
This yielded an exception of type `MissingMethodException` as the old assembly did not have the new methods. Was very surprised since static compilation succeeded without any issues. This means that the assembly optimization step is done after the compilation that can yield very nasty bugs as compile rules are not enforced.

**Comment by ltrzesniewski** [comment]
Here's a workaround I'm using, maybe it can help you too:

```xml
<!-- Work around https://github.com/dotnet/sdk/issues/12379 -->
<Target Name="DeleteR2RImages" BeforeTargets="_PrepareForReadyToRunCompilation" Condition="'$(PublishReadyToRun)' == 'true'">
  <PropertyGroup>
    <_R2RDirToRemove>$(IntermediateOutputPath)R2R</_R2RDirToRemove>
  </PropertyGroup>
  <RemoveDir Directories="$(_R2RDirToRemove)" />
</Target>
```

This basically clears the R2R cache before each publish, so it's not *great*, but it's still better than getting a broken build.

It works on the 3.1 SDK. I haven't tested it on .NET 5 yet but from a quick glance at the source, it should still work.

**Comment by jsheetzati** [comment]
This burned us once before when using R2R. Ended up w/ wrong nuget versions deployed to our Azure App Service. Afraid to re-enable the feature with this outstanding issue. 

Are we supposed to follow the steps outlined by @ltrzesniewski? I do not see any mention of this on the documentation. https://docs.microsoft.com/en-us/dotnet/core/deploying/ready-to-run

**Comment by alex-matei** [comment]
The workaround @ltrzesniewski made also works on .NET 5.

**Comment by ltrzesniewski** [comment]
This no longer reproduces with SDK 10.0.300, thanks @AntonLapounov 👍
